packages/transaction-controller/src/helpers/EtherscanRemoteTransactionSource.ts

```diff
@@ -51,20 +56,36 @@ export class EtherscanRemoteTransactionSource
   async fetchTransactions(
     request: RemoteTransactionSourceRequest,
   ): Promise<TransactionMeta[]> {
+    const releaseLock = await this.#mutex.acquire();
+    const acquiredTime = Date.now();
+
     const etherscanRequest: EtherscanTransactionRequest = {
       ...request,
       chainId: request.currentChainId,
     };
 
-    const transactions = this.#isTokenRequestPending
-      ? await this.#fetchTokenTransactions(request, etherscanRequest)
-      : await this.#fetchNormalTransactions(request, etherscanRequest);
+    try {
+      const transactions = this.#isTokenRequestPending
+        ? await this.#fetchTokenTransactions(request, etherscanRequest)
+        : await this.#fetchNormalTransactions(request, etherscanRequest);
 
-    if (this.#includeTokenTransfers) {
-      this.#isTokenRequestPending = !this.#isTokenRequestPending;
-    }
+      if (this.#includeTokenTransfers) {
+        this.#isTokenRequestPending = !this.#isTokenRequestPending;
+      }
 
-    return transactions;
+      return transactions;
+    } finally {
+      const elapsedTime = Date.now() - acquiredTime;
+      const remainingTime = Math.max(
+        0,
+        this.ETHERSCAN_RATE_LIMIT_INTERVAL - elapsedTime,
+      );
+      // Wait for the remaining time if it hasn't been 5 seconds yet
+      if (remainingTime > 0) {
+        await new Promise((resolve) => setTimeout(resolve, remainingTime));
```

Comment:
Rather than first awaiting the remaining time, should we release the lock in the callback of the `setTimeout`?

That way we don't block the return of our transactions to the caller, but can still hold other requests until the timeout is elapsed?

Reply:
I've got this one

Reply:
addressed [here](https://github.com/MetaMask/core/pull/3875)